drivers/network/tcpip/tcpip/icmp.c

```diff
@@ -0,0 +1,389 @@
+
+
+KDPC TimeoutDpc;
+KTIMER TimeoutTimer;
+static volatile INT16 IcmpSequence = 0;
+
+static
+UINT32 GetReplyStatus(PICMP_HEADER IcmpHeader)
+{
+    switch (IcmpHeader->Type)
+    {
+        case ICMP_TYPE_ECHO_REPLY:
+            return IP_SUCCESS;
+        case ICMP_TYPE_DEST_UNREACH:
+            switch (IcmpHeader->Code)
+            {
+                case ICMP_CODE_DU_NET_UNREACH:
+                    return IP_DEST_NET_UNREACHABLE;
+                case ICMP_CODE_DU_HOST_UNREACH:
+                    return IP_DEST_HOST_UNREACHABLE;
+                case ICMP_CODE_DU_PROTOCOL_UNREACH:
+                    return IP_DEST_PROT_UNREACHABLE;
+                case ICMP_CODE_DU_PORT_UNREACH:
+                    return IP_DEST_PORT_UNREACHABLE;
+                case ICMP_CODE_DU_FRAG_DF_SET:
+                    return IP_DEST_NET_UNREACHABLE;
+                case ICMP_CODE_DU_SOURCE_ROUTE_FAILED:
+                    return IP_BAD_ROUTE;
+                default:
+                    return IP_DEST_NET_UNREACHABLE;
+            }
+        case ICMP_TYPE_SOURCE_QUENCH:
+            return IP_SOURCE_QUENCH;
+        case ICMP_TYPE_TIME_EXCEEDED:
+            if (IcmpHeader->Code == ICMP_CODE_TE_REASSEMBLY)
+                return IP_TTL_EXPIRED_REASSEM;
+            else
+                return IP_TTL_EXPIRED_TRANSIT;
+        case ICMP_TYPE_PARAMETER:
+            return IP_PARAM_PROBLEM;
+        default:
+            return IP_REQ_TIMED_OUT;
+    }
+}
+
+NTSTATUS
+NTAPI
+ReceiveDatagram(
+    _In_opt_ PVOID TdiEventContext,
+    _In_ LONG SourceAddressLength,
+    _In_reads_bytes_(SourceAddressLength) PVOID SourceAddress,
+    _In_ LONG OptionsLength,
+    _In_reads_bytes_opt_(OptionsLength) PVOID Options,
+    _In_ ULONG ReceiveDatagramFlags,
+    _In_ ULONG BytesIndicated,
+    _In_ ULONG BytesAvailable,
+    _Out_ ULONG *OutBytesTaken,
+    _In_ PVOID Tsdu,
+    _Out_opt_ PIRP *IoRequestPacket)
+{
+    PICMP_PACKET_CONTEXT Context = (PICMP_PACKET_CONTEXT)TdiEventContext;
+    KIRQL OldIrql;
+
+    KeWaitForSingleObject(&Context->InitializationFinishedEvent, Executive, KernelMode, FALSE, NULL);
+
+    TI_DbgPrint(DEBUG_ICMP, ("Received datagram Context: 0x%p\n", TdiEventContext));
+
+    KeAcquireSpinLock(&Context->ProcessingLock, &OldIrql);
+
+    if (SourceAddressLength != sizeof(IPAddr))
+    {
+        Context->ErrorStatus = STATUS_UNSUCCESSFUL;
+    }
+    else
+    {
+        UINT16 HeaderSize = sizeof(IPv4_HEADER) + OptionsLength;
+        INT32 DataSize = min(BytesAvailable, UINT16_MAX) - HeaderSize - sizeof(ICMP_HEADER);
+        PVOID DataBuffer = (PUCHAR)Tsdu + HeaderSize + sizeof(ICMP_HEADER);
+        PIPv4_HEADER Header = (PIPv4_HEADER)Tsdu;
+        INT64 CurrentTime = KeQueryPerformanceCounter(NULL).QuadPart;
+        UINT32 RoundTripTime = (CurrentTime - Context->StartTicks) / Context->TimerResolution;
+        PICMP_HEADER IcmpHeader = (PICMP_HEADER)((PUCHAR)Tsdu + HeaderSize);
```

Comment:
I'd assume you would have to validate that the header is actually complete (i.e. `DataSize` is > 0) before accessing its data